### PR TITLE
[MDCT-2612] -Update validation for MLR overlay page

### PR DESCRIPTION
**Pull request by keeysnc**
### Description
Validations for following MLR question fields.  

Ticket:https://qmacbis.atlassian.net/jira/software/c/projects/MDCT/boards/232?modal=detail&selectedIssue=MDCT-2612
Questions 1.3, 2.3, 3.1, and 3.4 on the MLR Overlay page should have the following validation
1. No text characters (including N/A)
2. No negative numbers
3. No Zero
4. Numbers 1 OR above only (inclusive of 1) (Positive Numbers)

Ticket https://qmacbis.atlassian.net/jira/software/c/projects/MDCT/boards/232?modal=detail&selectedIssue=MDCT-2613
Question 4.6.1 has a slightly different validation scheme
* No text characters (including N/A)
* No negative numbers
* Numbers zero or above only (inclusive of 0)

### Related ticket(s)
MDCT-2612
MDCT-2613

---
### How to test
1. Log into Medical Loss Ratio (MLR) Report
2. Create MLR Submission
3. Fill out "Information for Primary Contract"
4. 'Add Program Reporting Information'
6. Enter the MLR Reporting and fill out form
7. Test the following questions above based on the requirements

If one of the fields above is invalid, the following should happen:
1.  An inline error should be displayed
2. You should be able to save the form (with incorrect input) but unable to submit it. 

* videos a little lengthy
https://github.com/Enterprise-CMCS/macpro-mdct-mcr/assets/14514294/f953e724-987b-45c6-a6d0-02bb54d85d02

---
### Author checklist

- [x] I have performed a self-review of my code
- [x] I have added [thorough](https://bit.ly/3zPrxuZ) tests, if necessary
- [ ] I have updated relevant documentation, if necessary
---

_convert to a different template: [test → val](?expand=1&template=test-to-val-deployment.md)_ | _[val → prod](?expand=1&template=val-to-prod-deployment.md)_

### CODE REVIEW

File: services/app-api/utils/validation/completionSchemas.ts [hunk 1]
```diff
@@ -17,6 +17,7 @@ export const error = {
   INVALID_DATE: "Response must be a valid date",
   INVALID_END_DATE: "End date can't be before start date",
   INVALID_NUMBER: "Response must be a valid number",
+  NON_POSITIVE_NUMBER: "Response must be a positive number",
```
Comment (karla-vm) [inline services/app-api/utils/validation/completionSchemas.ts:20]: Looking good! Just a quick question about the phrasing, wondering if this is clear for users that 0 is not included. Would you mind asking Kim or Megan about that?

**Review by karla-vm — DISMISSED** [review]

File: services/app-api/utils/validation/completionSchemas.ts (hunk 1 above)
Comment (keeysnc) [reply services/app-api/utils/validation/completionSchemas.ts:20]: Good point, thanks ill check.

Comment (keeysnc) [reply services/app-api/utils/validation/completionSchemas.ts]: @karla-vm "Response must be a number greater than zero"  confirmed by Megan, Kim & DG

File: services/app-api/utils/validation/schemaMap.test.ts [hunk 2]
```diff
@@ -90,6 +102,11 @@ describe("Schemas", () => {
     testSchema(number(), badNumberTestCases, false);
   });
 
+  test("Evaluate Number Schema using number positive scheme", () => {
+    testSchema(numberPositive(), goodPositiveNumberTestCases, true);
+    testSchema(numberPositive(), badNumberTestCases, false);
```
Comment (benmartin-coforma) [inline services/app-api/utils/validation/schemaMap.test.ts:107]: Request: Could you add a test case for zero, and one for a negative number? For this schema and the other two.

File: services/ui-src/src/verbiage/errors.ts [hunk 3]
```diff
@@ -14,6 +14,7 @@ export const validationErrors = {
   INVALID_URL: "Response must be a valid hyperlink/URL",
   INVALID_DATE: "Response must be a valid date",
   INVALID_END_DATE: "End date can't be before start date",
+  NON_POSITIVE_NUMBER: "Response must be equal to or greater than one",
```
Comment (benmartin-coforma) [inline services/ui-src/src/verbiage/errors.ts:17]: Comment: It's slightly irksome that this message indicates the value must be at least one, whereas our validation actually checks that the value is greater than zero. If we ever apply this validation to a field where 0.5 would be acceptable, we will need to adjust the message somehow.

**Review by benmartin-coforma — DISMISSED** [review]
This is good work so far. I'm requesting some minor changes before we merge it in.

File: services/ui-src/src/verbiage/errors.ts (hunk 3 above)
Comment (keeysnc) [reply services/ui-src/src/verbiage/errors.ts:17]: @benmartin-coforma Yea reading it back it does sound a little more complicated than "greater than zero". And making another message for decimals may be overkill. I'll circle back to Megan, Kim & David and see if we can switch it.

Comment (keeysnc) [reply services/ui-src/src/verbiage/errors.ts:17]: @benmartin-coforma There good with '"Response must be a number greater than zero".'

File: services/app-api/utils/validation/schemaMap.ts [hunk 4]
```diff
@@ -50,6 +52,19 @@ export const number = () =>
     },
   });
 
+// NUMBER POSITIVE - Positive numbers only
+export const numberPositive = () =>
+  string()
+    .required(error.REQUIRED_GENERIC)
```
[the remark this answers is not in the document]
Comment (keeysnc) [reply services/app-api/utils/validation/schemaMap.ts:58]: @benmartin-coforma I see what you mean. Possibly, I can ask David about this.

File: services/app-api/utils/validation/schemaMap.test.ts (hunk 2 above)
Comment (keeysnc) [reply services/app-api/utils/validation/schemaMap.test.ts]: New test added.

File: services/app-api/utils/validation/schemaMap.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (keeysnc) [reply services/app-api/utils/validation/schemaMap.ts:58]: Disregarded after speaking with Ben.

[the remark this answers is not in the document]
Comment (braxex) [reply services/app-api/utils/validation/schemaMap.ts:58]: Yes, to add on here -- users need to be allowed to blank fields and save empty values. So blank values should trigger inline field validation errors on the frontend, but the backend should allow those values to pass through so they can be saved in the database.

**Review by braxex — CHANGES_REQUESTED** [review]
Blocking: There is one issue I see here. [MDCT-2612](https://qmacbis.atlassian.net/browse/MDCT-2612) and [MDCT-2613](https://qmacbis.atlassian.net/browse/MDCT-2613) are being treated as if they are the same ask from a validation standpoint but they should be separate. 

- 2613 asks for validation that `input >= 0` for field 4.6.1
- 2612 asks for validation that `input >= 1` for fields 1.3, 2.3, 3.1, 3.4

We should be adding two new validation schemas here instead of using the same one for both.

**Comment by keeysnc** [comment]
> Blocking: There is one issue I see here. [MDCT-2612](https://qmacbis.atlassian.net/browse/MDCT-2612) and [MDCT-2613](https://qmacbis.atlassian.net/browse/MDCT-2613) are being treated as if they are the same ask from a validation standpoint but they should be separate.
> 
> * 2613 asks for validation that `input >= 0` for field 4.6.1
> * 2612 asks for validation that `input >= 1` for fields 1.3, 2.3, 3.1, 3.4
> 
> We should be adding two new validation schemas here instead of using the same one for both.

@braxex Updated. I added another scheme called numberNotLessThanZero for 4.6.1. (updated test as well)

File: services/app-api/utils/validation/schemaMap.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (keeysnc) [reply services/app-api/utils/validation/schemaMap.ts:59]: Currently the updates consist of saving empty values but not allowing a final submission until those values are filled in.

File: services/ui-src/src/verbiage/errors.ts [hunk 5]
```diff
@@ -14,6 +14,8 @@ export const validationErrors = {
   INVALID_URL: "Response must be a valid hyperlink/URL",
   INVALID_DATE: "Response must be a valid date",
   INVALID_END_DATE: "End date can't be before start date",
+  NUMBER_LESS_THAN_ONE: "Response must be equal to or greater than one",
+  NUMBER_LESS_THAN_ZERO: "Response must be equal to or greater than zero",
```
Comment (ntsummers1) [inline services/ui-src/src/verbiage/errors.ts:18]: Should this be 
```suggestion
  NUMBER_GREATER_THAN_ONE: "Response must be equal to or greater than one",
  NUMBER_GREATER_THAN_ZERO: "Response must be equal to or greater than zero",
```

Comment (ntsummers1) [reply services/ui-src/src/verbiage/errors.ts:18]: Or NUMBER_NOT_LESS_THAN_ONE to follow the above naming scheme

Comment (keeysnc) [reply services/ui-src/src/verbiage/errors.ts:18]: @ntsummers1 This brings up a good question, I wanted to write it as an error so when its used in situations like `message: error.NUMBER_LESS_THAN_ONE,` we know whats wrong with the input.

**Review by ntsummers1 — DISMISSED** [review]

**Review by ajaitasaini — APPROVED** [review]
amazing work with the PR + addressing all the feedback!!

File: services/app-api/utils/validation/completionSchemas.ts [hunk 6]
```diff
@@ -16,6 +16,8 @@ export const error = {
   INVALID_URL: "Response must be a valid hyperlink/URL",
   INVALID_DATE: "Response must be a valid date",
   INVALID_END_DATE: "End date can't be before start date",
+  NUMBER_LESS_THAN_ZERO: "Response must be equal to or greater than zero",
+  NUMBER_LESS_THAN_ONE: "Response must be equal to or greater than one",
```
Comment (braxex) [inline services/app-api/utils/validation/completionSchemas.ts:20]: Nit: This wording sounds weird to me. Should it be "Response must be greater than or equal to x" instead?

**Review by braxex — APPROVED** [review]
My blocker has been resolved. Thank you! 

I haven't pulled it down and run it locally, so we should be sure someone else does that pre-merge. Also you'll be blocked from merging until all conversations are resolved, so make sure you do that first. 👍

**Review by karla-vm — APPROVED** [review]
Looks good!!
